pkg/action/upgrade.go

```
@@ -0,0 +1,285 @@
+		statePart, err = utils.GetFullDeviceByLabel(u.Config.Runner, u.Config.RecoveryLabel, 5)
+		if err != nil {
+			return err
+		}
+	} else {
+		statePart, err = utils.GetFullDeviceByLabel(u.Config.Runner, u.Config.StateLabel, 5)
+		if err != nil {
+			return err
+		}
+	}
+
+	u.Info("Mounting state partition %s in %s", statePart.Path, upgradeStateDir)
+	// Create the dir only if not on recovery as on recovery the dir exists already
+	if !u.Config.RecoveryUpgrade {
+		err = u.Config.Fs.MkdirAll(upgradeStateDir, os.ModeDir)
+
+		if err != nil {
+			u.Error("Error creating statedir %s: %s", upgradeStateDir, err)
+			return err
+		}
+	}
+
+	err = u.Config.Mounter.Mount(statePart.Path, upgradeStateDir, statePart.FS, []string{"remount", "rw"})
+	if err != nil {
+		u.Error("Error mounting %s: %s", upgradeStateDir, err)
+
+		return err
+	}
+	defer func() {
+		if u.Config.RecoveryUpgrade {
+			// remount cos-state RO again
+			u.Info("Remounting %s as RO", upgradeStateDir)
+			err = u.Config.Mounter.Mount(statePart.Path, upgradeStateDir, statePart.FS, []string{"remount", "ro"})
+			if err != nil {
+				u.Error("Error remounting %s: %s", upgradeStateDir, err)
+			}
+		}
+	}()
+
+	// create transition.img
+	img := v1.Image{
+		File:       transitionImg,
+		Label:      strings.ToUpper(upgradeTarget),
+		Size:       u.Config.ImgSize,
+		FS:         constants.LinuxImgFs,
+		MountPoint: constants.UpgradeTempDir,
+		RootTree:   upgradeSource.source, // if source is a dir it will copy from here, if it's a docker img it uses Config.DockerImg IN THAT ORDER!
+	}
+
+	ele := elemental.NewElemental(u.Config)
+
+	if !u.Config.RecoveryUpgrade {
+		err = ele.CreateFileSystemImage(img)
+		if err != nil {
+			u.Error("Failed to create %s img: %s", transitionImg, err)
+			return err
+		}
+
+		// mount the transition img on targetDir, so we can install the upgraded files into targetDir, and they end up on the img
+		err = ele.MountImage(&img, []string{"rw"}...)
+		defer ele.UnmountImage(&img)
```

note this swallows unmounting errors. I think we should consider those, there could be errors on unmounting and having some explicit failure for that I think it is really good for debugging. This can be done in a follow up though. The install action uses kind of a simple (but verbose) pattern for that.

---

to be fair, do we really care in the upgrade? I mean, the next step after upgrade should be a reboot in which this will be unmounted automatically no? Plus the transition img will be autoremoved after a reboot. Its more of a "lets leave things as they were" than a "hard cleanup because we dont want to pollute the system". The magic of read only FS :laughing: